common/stats/stats.go

```diff
@@ -113,3 +113,75 @@ func (r *emaFixedWindowQPSTracker) ReportCounter(delta int64) {
 func (r *emaFixedWindowQPSTracker) QPS() float64 {
 	return r.qps.Load()
 }
+
+type (
+	bucket struct {
+		counter   int64
+		timeIndex int
+	}
+	rollingWindowQPSTracker struct {
+		sync.RWMutex
+		timeSource     clock.TimeSource
+		bucketInterval time.Duration
+		buckets        []bucket
+
+		counter   int64
+		timeIndex int
+	}
+)
+
+func NewRollingWindowQPSTracker(timeSource clock.TimeSource, bucketInterval time.Duration, numBuckets int) QPSTracker {
+	return &rollingWindowQPSTracker{
+		timeSource:     timeSource,
+		bucketInterval: bucketInterval,
+		buckets:        make([]bucket, numBuckets),
+	}
+}
+
+func (r *rollingWindowQPSTracker) Start() {
+}
+
+func (r *rollingWindowQPSTracker) Stop() {
+}
+
+func (r *rollingWindowQPSTracker) getCurrentTimeIndex() int {
+	now := r.timeSource.Now()
+	return int(now.UnixNano() / int64(r.bucketInterval))
+}
+
+func (r *rollingWindowQPSTracker) ReportCounter(delta int64) {
+	r.Lock()
+	defer r.Unlock()
```

Comment:
lock performance may not be ideal for high qps use cases. Try using atomic variables. Below is code snippet from chatpgt which might provides more insights.
```suggestion
// RollingWindow represents a lock-free rolling window
type RollingWindow struct {
	window      []uint64      // Atomic counters for each time bucket
	size        int           // Size of the window in seconds or intervals
	timeWindow  time.Duration // Time window duration (e.g., 1 second per bucket)
	startTime   uint64        // Start time of the window (in seconds since epoch)
}

// NewRollingWindow initializes a new rolling window
func NewRollingWindow(size int, timeWindow time.Duration) *RollingWindow {
	return &RollingWindow{
		window:     make([]uint64, size),
		size:       size,
		timeWindow: timeWindow,
		startTime:  uint64(time.Now().Unix()),
	}
}

// currentBucket calculates the current bucket based on the elapsed time
func (rw *RollingWindow) currentBucket() int {
	now := uint64(time.Now().Unix())
	elapsed := now - atomic.LoadUint64(&rw.startTime)
	return int(elapsed % uint64(rw.size))
}

// Increment increments the current time bucket
func (rw *RollingWindow) Increment() {
	bucket := rw.currentBucket()
	atomic.AddUint64(&rw.window[bucket], 1)
}

// GetQPS returns the total number of requests in the rolling window
func (rw *RollingWindow) GetQPS() int {
	now := uint64(time.Now().Unix())
	currentBucket := now % uint64(rw.size)
	total := 0

	for i := 0; i < rw.size; i++ {
		if uint64(i) == currentBucket {
			// Reset the current bucket since we're at the beginning of a new second
			atomic.StoreUint64(&rw.window[i], 0)
		} else {
			total += int(atomic.LoadUint64(&rw.window[i]))
		}
	}

	return total
}
```

Reply:
There are more than 2 variables needs to be touched and condition check is involved. The operation cannot be simplified with CAS.